samples/bluetooth/fast_pair/locator_tag/src/motion_detector/platform_thingy.c

```diff
@@ -0,0 +1,246 @@
+
+	__ASSERT_NO_MSG(!motion_detector.active);
+	LOG_INF("FMDN: motion detector started");
+	motion_detector.active = true;
+	k_sem_give(&gyro_poll_sem);
+	app_ui_state_change_indicate(APP_UI_STATE_MOTION_DETECTOR_ACTIVE, motion_detector.active);
+}
+
+static bool fmdn_motion_detector_period_expired(void)
+{
+	/* It is assumed that this function executes in the cooperative thread context. */
+	__ASSERT_NO_MSG(!k_is_preempt_thread());
+	__ASSERT_NO_MSG(!k_is_in_isr());
+
+	bool ret = motion_detected();
+
+	__ASSERT_NO_MSG(motion_detector.active);
+	LOG_INF("FMDN: motion detector period expired. Reporting that the motion was %sdetected",
+		ret ? "" : "not ");
+	memset(motion_detector.rot, 0, sizeof(motion_detector.rot));
+	return ret;
+}
+
+static void fmdn_motion_detector_stop(void)
+{
+	/* It is assumed that this function executes in the cooperative thread context. */
+	__ASSERT_NO_MSG(!k_is_preempt_thread());
+	__ASSERT_NO_MSG(!k_is_in_isr());
+
+	__ASSERT_NO_MSG(motion_detector.active);
+	LOG_INF("FMDN: motion detector stopped");
+	motion_detector.active = false;
+	memset(motion_detector.rot, 0, sizeof(motion_detector.rot));
+	app_ui_state_change_indicate(APP_UI_STATE_MOTION_DETECTOR_ACTIVE, motion_detector.active);
+}
+
+static const struct bt_fast_pair_fmdn_motion_detector_cb fmdn_motion_detector_cb = {
+	.start = fmdn_motion_detector_start,
+	.period_expired = fmdn_motion_detector_period_expired,
+	.stop = fmdn_motion_detector_stop,
+};
+
+static int sensor_configure(const struct device *sensor)
+{
+	struct sensor_value scale;
+	struct sensor_value sampling_freq;
+	struct sensor_value oversampling;
+	int err;
+
+	SENSOR_VALUE_SCALE_1000_DEG_PER_SEC_SET(scale);
+	err = sensor_attr_set(sensor, SENSOR_CHAN_GYRO_XYZ, SENSOR_ATTR_FULL_SCALE, &scale);
+	if (err) {
+		return err;
+	}
+
+	SENSOR_VALUE_OVERSAMPLING_DISABLE(oversampling);
+	err = sensor_attr_set(sensor, SENSOR_CHAN_GYRO_XYZ, SENSOR_ATTR_OVERSAMPLING,
+			      &oversampling);
+	if (err) {
+		return err;
+	}
+
+	SENSOR_VALUE_SAMPLING_FREQ_25_HZ_SET(sampling_freq);
+	err = sensor_attr_set(sensor, SENSOR_CHAN_GYRO_XYZ,
+			      SENSOR_ATTR_SAMPLING_FREQUENCY,
+			      &sampling_freq);
+	if (err) {
+		return err;
+	}
+
+	return err;
+}
+
+static void sensor_fetch(void)
+{
+	struct sensor_value val[ARRAY_SIZE(motion_detector.rot)];
+
+	sensor_sample_fetch(motion_detector.dev);
+	sensor_channel_get(motion_detector.dev, SENSOR_CHAN_GYRO_XYZ, val);
+	for (size_t i = 0; i < ARRAY_SIZE(motion_detector.rot); i++) {
+		/* Rotation = Angular velocity * Time = Angular velocity / Sampling frequency. */
+		motion_detector.rot[i] += sensor_value_to_double(&val[i]) / GYRO_SAMPLING_FREQ;
+	}
+}
+
+static void gyro_poll_thread_fn(void *dummy0, void *dummy1, void *dummy2)
+{
+	ARG_UNUSED(dummy0);
+	ARG_UNUSED(dummy1);
+	ARG_UNUSED(dummy2);
+
+	while (true) {
+		k_spinlock_key_t key;
+
+		k_sem_take(&gyro_poll_sem, K_FOREVER);
+
+		key = k_spin_lock(&motion_detector.lock);
```

Comment:
what is the purpose of having a preemptive thread if we request the kernel to disable preempting it during its execution (from the `k_sem_take` sleep time to the `k_msleep` sleep time)? Isn't it better to define this thread as COOP? Or move the spinlock to the `sensor_fetch` part right before the for loop (if you believe that `sensor_sample_fetch` / `sensor_channel_get` are long enough and it makes sense to allow the kernel to preempt this thread during these functions execution)?

Reply:
You're there's currently no point in keeping the thread as preemptive. I can't move the lock to `sensor_fetch` as the `if (motion_detector.active) {` also need to be inside the lock. I will remove the lock and set the thread to coop.

Reply:
Or alternatively I could just do `sensor_sample_fetch` and `sensor_channel_get` under spinlock unconditionally and then under spinlock:
```
if (motion_detector.active) {
	for (size_t i = 0; i < ARRAY_SIZE(motion_detector.rot); i++) {
		              /* Rotation = Angular velocity * Time = Angular velocity / Sampling frequency. */
		              motion_detector.rot[i] += sensor_value_to_double(&val[i]) / GYRO_SAMPLING_FREQ;
	              }
	k_sem_give(&gyro_poll_sem);
}

Reply:
That second idea might be actually more clean.

Reply:
So the whole preemptive thread code would look like that:
```
while (true) {
        k_spinlock_key_t key;
        struct sensor_value val[ARRAY_SIZE(motion_detector.rot)];
        
        k_sem_take(&gyro_poll_sem, K_FOREVER);

	sensor_sample_fetch(motion_detector.dev);
	sensor_channel_get(motion_detector.dev, SENSOR_CHAN_GYRO_XYZ, val);
        
        key = k_spin_lock(&motion_detector.lock);
        if (motion_detector.active) {
                for (size_t i = 0; i < ARRAY_SIZE(motion_detector.rot); i++) {
                         /* Rotation = Angular velocity * Time = Angular velocity / Sampling frequency. */ 
                        motion_detector.rot[i] += sensor_value_to_double(&val[i]) / GYRO_SAMPLING_FREQ; } 
                        k_sem_give(&gyro_poll_sem); }
```